### PR TITLE
fix: invalidate cache in core.request.add_haeder and fix some calls

### PR DESCRIPTION
**Pull request by jiangfucheng**
### Description

Fixes https://github.com/apache/apisix/issues/8811

### Checklist

- [ ] I have explained the need for this PR and the problem it solves
- [ ] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [ ] I have updated the documentation to reflect this change
- [ ] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

### CODE REVIEW

File: apisix/plugins/proxy-rewrite.lua [hunk 1]
```diff
@@ -311,18 +311,18 @@ do
         for i = 1, field_cnt, 2 do
             local val = core.utils.resolve_var(hdr_op.add[i + 1], ctx.var)
             local header = hdr_op.add[i]
-            core.request.add_header(header, val)
+            core.request.add_header(ctx, header, val)
```
Comment (tokers) [inline apisix/plugins/proxy-rewrite.lua:314]: We may add a case to ensure the warning message won't be generated anymore.

Comment (jiangfucheng) [reply apisix/plugins/proxy-rewrite.lua:314]: Done

**Comment by jiangfucheng** [comment]
image
This test can pass on my local machine😳.

File: t/core/request.t [hunk 2]
```diff
@@ -439,3 +439,59 @@ while ($i <= 101) {
...
+            core.request.add_header(ctx, "test_header", "test")
+            local h = core.request.header(ctx, "test_header")
+            ngx.say(ctx.headers.test_header)
+            core.request.add_header(ctx, "test_header", "abc")
+            ngx.say(ctx.headers.test_header)
+        }
+    }
+--- response_body
+test
+nil
```
Comment (soulbird) [inline t/core/request.t:478]: Should be `abc` ？

Comment (jiangfucheng) [reply t/core/request.t:478]: `add_header()` method will invalidate cache in `ctx`, `header()` method will add cache to `ctx`, so first value is "test" and the second value is "nil"

**Comment by jiangfucheng** [comment]
@soulbird @spacewander @tokers PTAL, Thanks.

File: t/core/request.t (hunk 2 above)
Comment (soulbird) [reply t/core/request.t:477]: Why does `add_header()` invalidate the cache?

File: t/plugin/proxy-rewrite3.t [hunk 3]
```diff
@@ -455,3 +455,12 @@ passed
 GET /echo HTTP/1.1
 --- response_headers
 test: test_in_set
+
+
+
+=== TEST 21: deprecated calls test
```
Comment (tokers) [inline t/plugin/proxy-rewrite3.t:461]: We can merge this case to `TEST 20`?

File: t/core/request.t (hunk 2 above)
Comment (jiangfucheng) [reply t/core/request.t:477]: This is the feature of current PR. `header(ctx, header_name)` method maybe get wrong values from ctx cache if the `add_header` method does not invalidate the cache.

File: t/plugin/proxy-rewrite3.t (hunk 3 above)
Comment (jiangfucheng) [reply t/plugin/proxy-rewrite3.t:461]: I think each test should foucus on one function point, even if their trigger conditions are same.

File: apisix/core/request.lua [hunk 4]
```diff
@@ -143,14 +143,28 @@ function _M.set_header(ctx, header_name, header_value)
...
         error(err)
     end
 
     req_add_header(header_name, header_value)
+
+    if not ctx then
+        ctx = ngx.ctx.api_ctx
+    end
+
+    ctx.headers = {}
```
Comment (monkeyDluffy6017) [inline apisix/core/request.lua:167]: The invalidate cache operatio should follow the `set_header` function

**Review by soulbird — APPROVED** [review]

File: apisix/core/request.lua [hunk 5]
```diff
@@ -143,14 +143,35 @@ function _M.set_header(ctx, header_name, header_value)
     end
 end
 
-function _M.add_header(header_name, header_value)
+function _M.add_header(ctx, header_name, header_value)
```
Comment (monkeyDluffy6017) [inline apisix/core/request.lua:146]: The function `add_header` and `set_header` are almost identical, maybe we could extract a common function

Comment (spacewander) [reply apisix/core/request.lua:146]: The names follow the corresponding API in OpenResty.

File: apisix/core/request.lua [hunk 6]
```diff
@@ -143,14 +143,35 @@ function _M.set_header(ctx, header_name, header_value)
...
+        changed = a6_request.is_request_header_set()
+    end
+
+
     req_add_header(header_name, header_value)
+
+    if is_apisix_or and not changed then
+        a6_request.clear_request_header()
+        if ctx and ctx.headers then
+            ctx.headers[header_name] = header_value
```
Comment (spacewander) [inline apisix/core/request.lua:172]: We should append instead of overwriting it?

Comment (jiangfucheng) [reply apisix/core/request.lua]: Fixed

File: apisix/core/request.lua [hunk 7]
```diff
@@ -131,26 +128,36 @@ function _M.set_header(ctx, header_name, header_value)
...
     if is_apisix_or and not changed then
         -- if the headers are not changed before,
         -- we can only update part of the cache instead of invalidating the whole
         a6_request.clear_request_header()
         if ctx and ctx.headers then
-            ctx.headers[header_name] = header_value
+            if override or not ctx.headers[header_name] then
+                ctx.headers[header_name] = header_value
+            else
+                ctx.headers[header_name] = ctx.headers[header_name] .. header_value
```
Comment (monkeyDluffy6017) [inline apisix/core/request.lua:141]: Do you have test cases to cover this?

Comment (jiangfucheng) [reply apisix/core/request.lua:141]: The Test 15 in `request.t` can cover this.

Comment (monkeyDluffy6017) [reply apisix/core/request.lua:141]: I think it is not corrent.
request headers like this:
```
 Foo: foo
 Foo: bar
 Foo: baz
```
will be a lua table like this:
```
 {"foo", "bar", "baz"}
```

Comment (jiangfucheng) [reply apisix/core/request.lua:141]: Yes, I misunderstood the API of `add_header`, I will fix it later.

**Comment by jiangfucheng** [comment]
@spacewander @monkeyDluffy6017 @soulbird Please help me review this PR, thanks❤️.

File: apisix/core/request.lua [hunk 8]
```diff
@@ -131,26 +129,41 @@ function _M.set_header(ctx, header_name, header_value)
...
 end
 
-function _M.add_header(header_name, header_value)
-    local err
-    header_name, err = _validate_header_name(header_name)
-    if err then
-        error(err)
+function _M.set_header(ctx, header_name, header_value)
+    if type(ctx) == "string" then
+        log.warn("DEPRECATED: use set_header(ctx, header_name, header_value) instead")
```
Comment (spacewander) [inline apisix/core/request.lua:155]: We can move the DEPRECATED log into the common function, and show a different log according to the override parameter.

File: apisix/core/request.lua [hunk 9]
```diff
@@ -131,26 +129,41 @@ function _M.set_header(ctx, header_name, header_value)
         changed = a6_request.is_request_header_set()
     end
 
-    ngx.req.set_header(header_name, header_value)
+    update_func(header_name, header_value)
```
Comment (spacewander) [inline apisix/core/request.lua:132]: We can use override instead of passing an extra function.

File: t/core/request.t [hunk 10]
```diff
@@ -439,3 +439,47 @@ while ($i <= 101) {
...
+        content_by_lua_block {
+            local core = require("apisix.core")
+            ngx.ctx.api_ctx = {}
+            local ctx = ngx.ctx.api_ctx
+            core.request.add_header(ctx, "test_header", "test")
+            local h = core.request.header(ctx, "test_header")
+            ngx.say(h)
+            core.request.add_header(ctx, "test_header", "t2")
+            local h2 = core.request.header(ctx, "test_header")
+            ngx.say("[" .. table.concat(h2, ", ") .. "]")
```
Comment (monkeyDluffy6017) [inline t/core/request.t:457]: is it better to use `core.json.encode` instead?

Comment (jiangfucheng) [reply t/core/request.t]: done

File: apisix/core/request.lua (hunk 8 above)
Comment (jiangfucheng) [reply apisix/core/request.lua]: Fixed.

File: apisix/core/request.lua (hunk 9 above)
Comment (jiangfucheng) [reply apisix/core/request.lua]: Fixed.

File: t/core/request.t [hunk 11]
```diff
@@ -439,3 +439,47 @@ while ($i <= 101) {
...
 --- response_body
 101
+
+
+
+=== TEST 15: add header
+--- config
+    location /t {
+        content_by_lua_block {
+            local core = require("apisix.core")
```
Comment (soulbird) [inline t/core/request.t:449]: It is better to use `local json = require("toolkit.json")`, because it will sort the keys of the table

Comment (jiangfucheng) [reply t/core/request.t:449]: Thanks, fixed.

File: t/plugin/proxy-rewrite3.t [hunk 12]
```diff
@@ -619,3 +622,50 @@ GET /test/plugin/proxy/rewrite HTTP/1.1
     }
 --- response_body
 /plugin_proxy_rewrite?a=c
+
+
+
+=== TEST 27: set route
```
Comment (soulbird) [inline t/plugin/proxy-rewrite3.t:628]: Is it a duplicate of test 19?

Comment (jiangfucheng) [reply t/plugin/proxy-rewrite3.t:628]: Just want prepare configuration for Test 28.

Comment (soulbird) [reply t/plugin/proxy-rewrite3.t:628]: Aren't Test 28 and Test 20 duplicates?

Comment (jiangfucheng) [reply t/plugin/proxy-rewrite3.t:628]: Ok, I will merge them. At first, I just want one test focus on one function pont.

Comment (jiangfucheng) [reply t/plugin/proxy-rewrite3.t]: Fixed.

**Review by soulbird — DISMISSED** [review]

File: apisix/core/request.lua [hunk 13]
```diff
@@ -131,26 +135,37 @@ function _M.set_header(ctx, header_name, header_value)
         changed = a6_request.is_request_header_set()
     end
 
-    ngx.req.set_header(header_name, header_value)
+    if override then
+        ngx.req.set_header(header_name, header_value)
```
Comment (spacewander) [inline apisix/core/request.lua:139]: For symmetry beauty, let's also localize this function too.

Comment (jiangfucheng) [reply apisix/core/request.lua]: Fixd.

**Review by spacewander — APPROVED** [review]

**Review by monkeyDluffy6017 — APPROVED** [review]

**Review by soulbird — APPROVED** [review]
